### PR TITLE
Partial fetch

### PR DESCRIPTION
**Pull request by michelvocks**
Could use some tests but should be ready for review.

### CODE REVIEW

File: provider/schema/execution_test.go [hunk 1]
```diff
@@ -316,6 +316,26 @@ func TestExecutionData_ResolveTable(t *testing.T) {
 		assert.Equal(t, []string{"name", "name_no_prefix", "prefix_name", "cq_id", "meta", "injected_field"}, expectedResource.columns)
 		assert.Equal(t, []interface{}{"other", "name_no_prefix", "prefix_name", expectedResource.cqId, expectedResource.Get("meta"), 1}, values)
 	})
+
+	t.Run("test partial fetch", func(t *testing.T) {
```
Comment (roneli) [inline provider/schema/execution_test.go:320]: I think this is a great test but move moving parts can fail in a fetch as stated above

**Review by roneli — CHANGES_REQUESTED** [review]
Found some edge cases we missed, I really liked the implementation

File: provider/schema/execution.go [hunk 2]
```diff
@@ -127,7 +170,7 @@ func (e ExecutionData) callTableResolve(ctx context.Context, client ClientMeta,
 	// check if channel iteration stopped because of resolver failure
 	if resolverErr != nil {
 		client.Logger().Error("received resolve resources error", "table", e.Table.Name, "error", resolverErr)
```
[the remark this answers is not in the document]
Comment (michelvocks) [reply provider/schema/execution.go:172]: I disagree. I think partial fetching shouldn't change the severity of log messages since a log message severity is an indicator of the issue severity (like the name says). Partial fetching doesn't solve or reduces the severity of issues but instead it only skips certain issues to make sure we that is get the most out of the run.

File: provider/schema/execution.go [hunk 3]
```diff
@@ -136,7 +179,7 @@ func (e ExecutionData) callTableResolve(ctx context.Context, client ClientMeta,
 	return nc, nil
 }
 
-func (e ExecutionData) resolveResources(ctx context.Context, meta ClientMeta, parent *Resource, objects []interface{}) error {
+func (e *ExecutionData) resolveResources(ctx context.Context, meta ClientMeta, parent *Resource, objects []interface{}) error {
```
[the remark this answers is not in the document]
Comment (michelvocks) [reply provider/schema/execution.go:182]: Agreed however I feel like this might be major refactor to the `e.DB.Insert` function. Would like to do that separately to now blow up this PR.

File: provider/schema/execution.go [hunk 4]
```diff
@@ -159,25 +202,32 @@ func (e ExecutionData) resolveResources(ctx context.Context, meta ClientMeta, pa
...
 	}
 	return nil
 }
 
-func (e ExecutionData) resolveResourceValues(ctx context.Context, meta ClientMeta, resource *Resource) (err error) {
+func (e *ExecutionData) resolveResourceValues(ctx context.Context, meta ClientMeta, resource *Resource) (err error) {
 	defer func() {
 		if r := recover(); r != nil {
 			e.Logger.Error("resolve resource recovered from panic", "table", e.Table.Name, "stack", string(debug.Stack()))
 			err = fmt.Errorf("failed resolve resource. Error: %s", r)
```
[the remark this answers is not in the document]
Comment (michelvocks) [reply provider/schema/execution.go:225]: Makes sense. Thanks!

File: provider/schema/execution.go [hunk 5]
```diff
@@ -159,25 +202,32 @@ func (e ExecutionData) resolveResources(ctx context.Context, meta ClientMeta, pa
...
+			if partialFetchErr := e.checkPartialFetchError(err, resource, "resolve resource recovered from panic"); partialFetchErr != nil {
+				err = partialFetchErr
+			} else {
+				err = nil
+			}
 		}
 	}()
 	if err = e.resolveColumns(ctx, meta, resource, resource.table.Columns); err != nil {
```
[the remark this answers is not in the document]
Comment (michelvocks) [reply provider/schema/execution.go:233]: Thanks!

File: provider/schema/execution.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (roneli) [reply provider/schema/execution.go:182]: Inserting a resource is part of the fetch, I agree, we can do this in a separate PR, but many of the "errors" come from duplicate key etc' of 1 value out of 1000 inserted resources, would be a shame to fail on 999 of them.

File: provider/schema/execution.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (roneli) [reply provider/schema/execution.go:172]: I agree, but, wouldn't that be confusing log wise, its a so called "good red" in your log, when people see errors its usually alarming.

**Review by roneli — APPROVED** [review]
looking good, we just need to address the two final comments I believe, I will have another look at it after and we will be able to merge the PR.

File: provider/schema/execution.go [hunk 6]
```diff
@@ -150,55 +195,95 @@ func (e ExecutionData) resolveResources(ctx context.Context, meta ClientMeta, pa
...
 			}
 		}
 	}
 	return nil
 }
 
-func (e ExecutionData) resolveResourceValues(ctx context.Context, meta ClientMeta, resource *Resource) (err error) {
+func (e *ExecutionData) copyDataIntoDB(ctx context.Context, resources Resources, shouldCascade bool) (Resources, error) {
+	partialFetchErrorOccurred := false
+	if err := e.Db.CopyFrom(ctx, resources, shouldCascade, e.extraFields); err != nil {
```
Comment (roneli) [inline provider/schema/execution.go:222]: invert this, so if copyfrom doesn't fail, return nil. if bulk insert doesn't fail return nil. if we are not partial fetch return err, insert one by one

**Review by roneli — CHANGES_REQUESTED** [review]
Looks good, minor comment on logic flow

File: provider/schema/execution.go (hunk 6 above)
Comment (michelvocks) [reply provider/schema/execution.go]: Thanks!

**Review by roneli — APPROVED** [review]
LGTM
